README.md
# metalsmith-htaccess

[![npm Version](https://badgen.net/npm/v/metalsmith-htaccess?icon=npm)](https://www.npmjs.com/package/metalsmith-htaccess)
[![npm Weekly Downloads](https://badgen.net/npm/dw/metalsmith-htaccess)](https://www.npmjs.com/package/metalsmith-htaccess)

[![Known Vulnerabilities](https://snyk.io/test/npm/metalsmith-htaccess/badge.svg)](https://snyk.io/test/npm/metalsmith-htaccess)
[![Test Coverage](https://badgen.net/codecov/c/github/emmercm/metalsmith-htaccess/master?icon=codecov)](https://codecov.io/gh/emmercm/metalsmith-htaccess)
[![Maintainability](https://badgen.net/codeclimate/maintainability/emmercm/metalsmith-htaccess?icon=codeclimate)](https://codeclimate.com/github/emmercm/metalsmith-htaccess/maintainability)

[![GitHub](https://badgen.net/badge/emmercm/metalsmith-htaccess/purple?icon=github)](https://github.com/emmercm/metalsmith-htaccess)
[![License](https://badgen.net/github/license/emmercm/metalsmith-htaccess?color=grey)](https://github.com/emmercm/metalsmith-htaccess/blob/master/LICENSE)

A Metalsmith plugin to create an Apache HTTP Server `.htaccess` configuration file.

## Installation

```bash
npm install --save metalsmith-htaccess
```

## JavaScript Usage

```javascript
const Metalsmith = require('metalsmith');
const htaccess   = require('metalsmith-htaccess');

Metalsmith(__dirname)
    .use(htaccess({
        // options here
    }))
    .build((err) => {
        if (err) {
            throw err;
        }
    });
```

## Options

### Core

#### `core.defaultCharset` (optional)

Type: `string` Default: `utf-8`

Sets [`AddDefaultCharset`](https://httpd.apache.org/docs/current/mod/core.html#adddefaultcharset).

#### `core.serverSignature` (optional)

Type: `boolean` Default: `false`

Enables or disables [`ServerSignature`](https://httpd.apache.org/docs/current/mod/core.html#serversignature).

#### `core.errorDocuments` (optional)

Type: `object`

An object of [`ErrorDocument`](https://httpd.apache.org/docs/current/mod/core.html#errordocument)s, example:

```json
{
    "core": {
        "errorDocuments": {
            "401": "/error_pages/401.html",
            "404": "/error_pages/404.html",
            "500": "/error_pages/500.html"
        }
    }
}
```

### Core Options

#### `core.options.followSymlinks` (optional)

Type: `boolean` Default: `true`

Enables or disables [`FollowSymlinks`](https://httpd.apache.org/docs/current/mod/core.html#options).

#### `core.options.indexes` (optional)

Type: `boolean` Default: `false`

Enables or disables [`Indexes`](https://httpd.apache.org/docs/current/mod/core.html#options).

#### `core.options.multiViews` (optional)

Type: `boolean` Default: `false`

Enables or disables [`MultiViews`](https://httpd.apache.org/docs/current/mod/core.html#options).

### Deflate

#### `deflate.mimeTypes` (optional)

Type: `string[]` Default: `["image/svg+xml", "application/javascript", "application/rss+xml", "application/vnd.ms-fontobject", "application/x-font", "application/x-font-opentype", "application/x-font-otf", "application/x-font-truetype", "application/x-font-ttf", "application/x-javascript", "application/xhtml+xml", "application/xml", "font/opentype", "font/otf", "font/ttf", "image/x-icon", "text/css", "text/html", "text/javascript", "text/plain", "text/xml"]`

A list of MIME types to [`AddOutputFilterByType DEFLATE`](https://httpd.apache.org/docs/current/mod/mod_deflate.html#enable).

### Dir

#### `dir.index` (optional)

Type: `string` Default: `index.html index.htm index.php index.cgi`

Set [`DirectoryIndex`](https://httpd.apache.org/docs/current/mod/mod_dir.html#directoryindex).

### Environment

#### `environment.serverAdminEmail` (optional)

Type: `string`

Set [`SetEnv SERVER_ADMIN`](https://httpd.apache.org/docs/current/mod/mod_env.html#setenv).

#### `environment.timezone` (optional)

Type: `string`

Set [`SetEnv TZ`](https://httpd.apache.org/docs/current/mod/mod_env.html#setenv).

### Expires

#### `expires.default` (optional)

Type: `string` Default: `access plus 2 days`

Set [`ExpiresDefault`](https://httpd.apache.org/docs/current/mod/mod_expires.html#expiresdefault).

#### `expires.types` (optional)

Type: `object` Default:

```json
{
    "image/jpg": "access plus 1 month",
    "image/svg+xml": "access 1 month",
    "image/gif": "access plus 1 month",
    "image/jpeg": "access plus 1 month",
    "image/png": "access plus 1 month",
    "text/css": "access plus 1 month",
    "text/javascript": "access plus 1 month",
    "application/javascript": "access plus 1 month",
    "application/x-shockwave-flash": "access plus 1 month",
    "image/ico": "access plus 1 month",
    "image/x-icon": "access plus 1 month",
    "text/html": "access plus 600 seconds"
}
```

Set [`ExpiresByType`](https://httpd.apache.org/docs/current/mod/mod_expires.html#expiresbytype).

### Gzip

#### `gzip.canNegotiate` (optional)

Type: `boolean`

Enables or disables `mod_gzip_can_negotiate`.

#### `gzip.dechunk` (optional)

Type: `boolean`

Enables or disables `mod_gzip_dechunk`.

#### `gzip.include` (optional)

Type: `string[]`

A list of `mod_gzip_item_include`.

#### `gzip.exclude` (optional)

Type: `string[]`

A list of `mod_gzip_item_exclude`.

### Headers

#### `headers.etag` (optional)

Type: `boolean` Default: `false`

Enables or disables [ETag header](https://httpd.apache.org/docs/current/mod/core.html#fileetag).

### MIME

#### `mime.defaultLanguage` (optional)

Type: `string`

Sets [`DefaultLanguage`](https://httpd.apache.org/docs/current/mod/mod_mime.html#defaultlanguage).

#### `mime.languages` (optional)

Type: `object`

An object of [`AddLanguage`](https://httpd.apache.org/docs/current/mod/mod_mime.html#addlanguage)s, example:

```json
{
    "mime": {
        "languages": {
            "en": "en",
            "en-gb": ".en",
            "en-us": ".en"
        }
    }
}
```

#### `mime.charsets` (optional)

Type: `object`

An object of [`AddCharset`](https://httpd.apache.org/docs/current/mod/mod_mime.html#addcharset)s, example:

```json
{
    "mime": {
        "charsets": {
            "EUC-JP": ".euc",
            "ISO-2022-JP": ".jis",
            "SHIFT_JIS": ".sjis"
        }
    }
}
```

#### `mime.types` (optional)

Type: `object`

An object of [`AddType`](https://httpd.apache.org/docs/current/mod/mod_mime.html#addtype)s, example:

```json
{
    "mime": {
        "types": {
            "image/gif": ".gif",
            "image/jpeg": "jpeg jpg jpe"
        }
    }
}
```

### Rewrite

#### `rewrite.404` (optional)

Type: `string`

Sets the [404 redirect page](https://httpd.apache.org/docs/current/mod/mod_rewrite.html).

#### `rewrite.options` (optional)

Type: `string` Default: `Inherit`

Sets [`RewriteOptions`](https://httpd.apache.org/docs/current/mod/mod_rewrite.html#rewriteoptions).

#### `rewrite.allowedMethods` (optional)

Type: `string[]`

A list of [allowed HTTP methods](https://httpd.apache.org/docs/current/mod/mod_rewrite.html#rewriteoptions).

#### `rewrite.httpsRedirect` (optional)

Type: `boolean`

Enable [HTTPS redirection](https://httpd.apache.org/docs/current/mod/mod_rewrite.html#rewriteoptions).

### Spelling

#### `spelling.check` (optional)

Type: `boolean`

Enables or disables [`CheckSpelling`](https://httpd.apache.org/docs/current/mod/mod_speling.html#checkspelling).

## Changelog

[Changelog](./CHANGELOG.md)


CHANGELOG.md
# Changelog

## v0.2.0 / 2020-06-30

- Set default for `expires.default`.
- Fixed wrong `<IfModule>` for DEFLATE.
- Changed `deflate.mimeTypes` to output only one line.
- Set default for `deflate.mimeTypes`.
- Added support for `ExpiresByType` (`expires.types`), with default.
- Removed gzip defaults.
- Removed `[R=302]` from `rewrite.404` `RewriteRule`.
- Added `[NE]` to `rewrite.httpsRedirect` `RewriteRule`.

## v0.1.0 / 2019-05-31

- Added Jest tests.
- `rewrite.404` now uses an HTTP 302 instead of 307.
- Falsy data in dictionaries and arrays is now filtered out.

## v0.0.2 / 2019-05-11

- Fixed option defaulting.
- Fixed linting.
- Created README.
- Created CHANGELOG.
- Updated dependencies:
  - `eslint`

## v0.0.1 / 2019-05-10

- Initial version.


lib/index.test.js
'use strict';

const {
  existsSync, mkdirSync, readdirSync, readFileSync, statSync,
} = require('fs');
const { join } = require('path');

const Metalsmith = require('metalsmith');
const assertDir = require('assert-dir-equal');

const htaccess = require('./index');

const test = (dir, config) => {
  describe(dir, () => {
    // Allow src directory to not exist / be empty and not committed
    if (!existsSync(`${dir}/src`)) {
      mkdirSync(`${dir}/src`);
    }

    ['/', '\\'].forEach((pathSeparator) => {
      it(`should build with path separator "${pathSeparator}"`, (testDone) => {
        Metalsmith(`${dir}`)
          // Convert the path separators to the one being tested
          .use((files, metalsmith, done) => {
            Object.keys(files)
              .forEach((filename) => {
                const newFilename = filename.replace(/[/\\]/g, pathSeparator);
                if (newFilename !== filename) {
                  files[newFilename] = files[filename];
                  delete files[filename];
                }
              });
            done();
          })
          // Run the plugin
          .use(htaccess(config))
          // Convert the path separators back to system default
          .use((files, metalsmith, done) => {
            const properPathSeparator = process.platform === 'win32' ? '\\' : '/';
            Object.keys(files)
              .forEach((filename) => {
                const newFilename = filename.replace(/[/\\]/g, properPathSeparator);
                if (newFilename !== filename) {
                  files[newFilename] = files[filename];
                  delete files[filename];
                }
              });
            done();
          })
          // Test the output
          .build((err) => {
            if (config.error) {
              expect(err)
                .toBe(config.error);
            } else {
              expect(err)
                .toBeNull();
            }

            if (err) {
              testDone();
              return;
            }

            assertDir(`${dir}/build`, `${dir}/expected`, { filter: () => true });
            testDone();
          });
      });
    });
  });
};

describe('metalsmith-htaccess', () => {
  const dirs = (p) => readdirSync(p)
    .map((f) => join(p, f))
    .filter((f) => statSync(f).isDirectory());
  dirs('lib/fixtures')
    .forEach((dir) => {
      const config = existsSync(`${dir}/config.json`) ? JSON.parse(readFileSync(`${dir}/config.json`).toString()) : {};
      test(dir, config);
    });
});


lib/index.js
'use strict';

const deepmerge = require('deepmerge');

const generate = (options) => {
  let htaccess = '';

  /* ***** Core ***** */

  htaccess += '<Files .htaccess>\n';
  htaccess += '\tOrder Allow,Deny\n';
  htaccess += '\tDeny from all\n';
  htaccess += '</Files>\n';

  if (options.core) {
    if (options.core.options) {
      if (options.core.options.followSymlinks) {
        htaccess += 'Options +FollowSymlinks\n';
      } else {
        htaccess += 'Options -FollowSymlinks\n';
      }

      if (options.core.options.indexes) {
        htaccess += 'Options +Indexes\n';
      } else {
        htaccess += 'Options -Indexes\n';
      }

      if (options.core.options.multiViews) {
        htaccess += 'Options +MultiViews\n';
      } else {
        htaccess += 'Options -MultiViews\n';
      }
    }

    if (options.core.defaultCharset) {
      htaccess += `AddDefaultCharset ${options.core.defaultCharset}\n`;
    }

    if (options.core.serverSignature) {
      htaccess += 'ServerSignature On\n';
    } else {
      htaccess += 'ServerSignature Off\n';
    }

    if (options.core.errorDocuments && Object.keys(options.core.errorDocuments).length) {
      Object.keys(options.core.errorDocuments)
        .forEach((code) => {
          htaccess += `ErrorDocument ${code} ${options.core.errorDocuments[code]}\n`;
        });
    }
  }

  if (options.headers && options.headers.etag) {
    htaccess += `FileETag ${options.headers.etag}\n`;
  } else {
    htaccess += 'FileETag none\n';
  }

  /* ***** mod_deflate ***** */

  if (options.deflate) {
    htaccess += '<IfModule mod_deflate.c>\n';

    if (options.deflate.mimeTypes && options.deflate.mimeTypes.length) {
      htaccess += `\tAddOutputFilterByType DEFLATE ${options.deflate.mimeTypes.join(' ')}\n`;
    }

    // https://gtmetrix.com/enable-gzip-compression.html
    htaccess += '\tBrowserMatch ^Mozilla/4 gzip-only-text/html\n';
    htaccess += '\tBrowserMatch ^Mozilla/4\\.0[678] no-gzip\n';
    htaccess += '\tBrowserMatch \\bMSIE !no-gzip !gzip-only-text/html\n';
    htaccess += '\tHeader append Vary User-Agent\n';

    htaccess += '</IfModule>\n';
  }

  /* ***** mod_dir ***** */

  if (options.dir) {
    htaccess += '<IfModule mod_dir.c>\n';

    if (options.dir.index) {
      htaccess += `\tDirectoryIndex ${options.dir.index}\n`;
    }

    htaccess += '</IfModule>\n';
  }

  /* ***** Environment ***** */

  if (options.environment) {
    if (options.environment.serverAdminEmail) {
      htaccess += `SetEnv SERVER_ADMIN ${options.environment.serverAdminEmail}\n`;
    }

    // https://www.inmotionhosting.com/support/website/general-server-setup/tz-ref-table
    if (options.environment.timezone) {
      htaccess += `SetEnv TZ ${options.environment.timezone}\n`;
    }
  }

  /* ***** mod_expires ***** */

  if (options.expires) {
    htaccess += '<IfModule mod_expires.c>\n';
    htaccess += '\tExpiresActive On\n';

    if (options.expires.default) {
      htaccess += `\tExpiresDefault "${options.expires.default}"\n`;
    }

    if (options.expires.types && Object.keys(options.expires.types).length) {
      Object.keys(options.expires.types)
        .forEach((type) => {
          const time = options.expires.types[type];
          htaccess += `\tExpiresByType ${type} "${time}"\n`;
        });
    }

    htaccess += '</IfModule>\n';
  }

  /* ***** mod_gzip ***** */

  if (options.gzip) {
    htaccess += '<IfModule mod_gzip.c>\n';
    htaccess += '\tmod_gzip_on Yes\n';

    if (options.gzip.canNegotiate) {
      htaccess += '\tmod_gzip_can_negotiate Yes\n';
    } else {
      htaccess += '\tmod_gzip_can_negotiate No\n';
    }

    if (options.gzip.dechunk) {
      htaccess += '\tmod_gzip_dechunk Yes\n';
    } else {
      htaccess += '\tmod_gzip_dechunk No\n';
    }

    if (options.gzip.include && options.gzip.include.length) {
      options.gzip.include
        .forEach((include) => {
          htaccess += `\tmod_gzip_item_include ${include}\n`;
        });
    }

    if (options.gzip.exclude && options.gzip.exclude.length) {
      options.gzip.exclude
        .forEach((exclude) => {
          htaccess += `\tmod_gzip_item_exclude ${exclude}\n`;
        });
    }

    htaccess += '</IfModule>\n';
  }

  /* ***** mod_headers ***** */

  if (options.headers) {
    htaccess += '<IfModule mod_headers.c>\n';

    // https://htaccessbook.com/disable-etags/
    if (options.headers.etag) {
      htaccess += `\tHeader set ETag ${options.headers.etag}\n`;
    } else {
      htaccess += '\tHeader unset ETag\n';
    }

    htaccess += '</IfModule>\n';
  }

  /* ***** mod_mime ***** */

  if (options.mime) {
    htaccess += '<IfModule mod_mime.c>\n';

    if (options.mime.defaultLanguage) {
      htaccess += `\tDefaultLanguage ${options.mime.defaultLanguage}\n`;
    }

    if (options.mime.languages && Object.keys(options.mime.languages).length) {
      Object.keys(options.mime.languages)
        .forEach((language) => {
          htaccess += `\tAddLanguage ${language} ${options.mime.languages[language]}\n`;
        });
    }

    if (options.mime.charsets && Object.keys(options.mime.charsets).length) {
      Object.keys(options.mime.charsets)
        .forEach((charset) => {
          htaccess += `\tAddCharset ${charset} ${options.mime.charsets[charset]}\n`;
        });
    }

    if (options.mime.types && Object.keys(options.mime.types).length) {
      Object.keys(options.mime.types)
        .forEach((type) => {
          htaccess += `\tAddType ${type} ${options.mime.types[type]}\n`;
        });
    }

    htaccess += '</IfModule>\n';
  }

  /* ***** mod_rewrite ***** */

  if (options.rewrite) {
    htaccess += '<IfModule mod_rewrite.c>\n';
    htaccess += '\tRewriteEngine on\n';

    if (options.rewrite.options) {
      htaccess += `\tRewriteOptions ${options.rewrite.options}\n`;
    }

    // https://htaccessbook.com/control-request-methods/
    if (options.rewrite.allowedMethods && options.rewrite.allowedMethods.length) {
      htaccess += `\tRewriteCond %{REQUEST_METHOD} !^(${options.rewrite.allowedMethods.join('|')}) [NC]\n`;
      htaccess += '\tRewriteRule (.*) - [F,L]\n';
    }

    if (options.rewrite['404']) {
      htaccess += '\tRewriteCond %{REQUEST_FILENAME} !-f\n';
      htaccess += '\tRewriteCond %{REQUEST_FILENAME} !-d\n';
      htaccess += `\tRewriteRule ^(.*)$ ${options.rewrite['404']} [L]\n`;
    }

    if (options.rewrite.httpsRedirect) {
      htaccess += '\tRewriteCond %{HTTPS} off\n';
      htaccess += '\tRewriteRule ^(.*)$ https://%{HTTP_HOST}%{REQUEST_URI} [L,R=301,NE]\n';
    }

    htaccess += '</IfModule>\n';
  }

  /* ***** mod_speling ***** */

  if (options.spelling) {
    // (Yes, it really is spelled this way)
    htaccess += '<IfModule mod_speling.c>\n';

    if (options.spelling.check) {
      htaccess += '\tCheckSpelling on\n';
    } else {
      htaccess += '\tCheckSpelling off\n';
    }

    htaccess += '</IfModule>\n';
  }

  /* ***** ***** ***** */

  return htaccess;
};

module.exports = (options) => {
  /**
   * Defaults are set with recommendations from the following websites:
   *  - https://htaccessbook.com/useful-htaccess-rules
   *  - https://perishablepress.com/stupid-htaccess-tricks
   *  - https://varvy.com/pagespeed/enable-compression.html
   *  - https://help.dreamhost.com/hc/en-us/articles/216363157-How-can-I-cache-my-site-with-an-htaccess-file-
   *  - https://help.dreamhost.com/hc/en-us/articles/215747758-Force-your-site-to-load-securely-with-an-htaccess-file
   */
  options = deepmerge({
    core: {
      options: {
        followSymlinks: true,
        indexes: false,
        multiViews: false,
      },
      defaultCharset: 'utf-8',
      serverSignature: false,
    },
    deflate: {
      mimeTypes: [
        'image/svg+xml',
        'application/javascript',
        'application/rss+xml',
        'application/vnd.ms-fontobject',
        'application/x-font',
        'application/x-font-opentype',
        'application/x-font-otf',
        'application/x-font-truetype',
        'application/x-font-ttf',
        'application/x-javascript',
        'application/xhtml+xml',
        'application/xml',
        'font/opentype',
        'font/otf',
        'font/ttf',
        'image/x-icon',
        'text/css',
        'text/html',
        'text/javascript',
        'text/plain',
        'text/xml',
      ],
    },
    dir: {
      index: 'index.html index.htm index.php index.cgi',
    },
    expires: {
      default: 'access plus 2 days',
      types: {
        'image/jpg': 'access plus 1 month',
        'image/svg+xml': 'access 1 month',
        'image/gif': 'access plus 1 month',
        'image/jpeg': 'access plus 1 month',
        'image/png': 'access plus 1 month',
        'text/css': 'access plus 1 month',
        'text/javascript': 'access plus 1 month',
        'application/javascript': 'access plus 1 month',
        'application/x-shockwave-flash': 'access plus 1 month',
        'image/ico': 'access plus 1 month',
        'image/x-icon': 'access plus 1 month',
        'text/html': 'access plus 600 seconds',
      },
    },
    headers: {
      etag: false,
    },
    rewrite: {
      options: 'Inherit',
    },
  }, options || {}, { arrayMerge: (destinationArray, sourceArray) => sourceArray });

  return (files, metalsmith, done) => {
    const htaccess = generate(options);

    files['.htaccess'] = {
      contents: Buffer.from(htaccess),
    };

    done();
  };
};
